Packs/ApiModules/Scripts/MicrosoftApiModule/MicrosoftApiModule.py

```diff
@@ -0,0 +1,329 @@
+        else:
+            token_retrieval_url = auth_id_and_token_retrieval_url[1]
+
+        return cls(tenant_id=tenant_id, auth_id=auth_id, enc_key=enc_key,  # type: ignore[misc]
+                   token_retrieval_url=token_retrieval_url, auth_type=OPROXY_AUTH_TYPE,
+                   app_name=app_name, refresh_token=refresh_token, *args, **kwargs)
+
+    @classmethod
+    def from_self_deployed(cls, tenant_id: str, client_id: str, client_secret: str, scope: str = '', resource: str = '',
+                           app_url: str = '', *args, **kwargs):
+        """
+        Args:
+            tenant_id: The self deployed tenant ID
+            client_id: The self deployed client ID
+            client_secret: The self deployed client secret
+            scope: The self deployed application scope
+            resource: The self deployed application resource
+            app_url: The self deployed application request URL
+        Returns:
+            An instance of Microsoft Client with self deployed application authentication.
+        """
+        return cls(tenant_id=tenant_id, client_id=client_id,  # type: ignore[misc]
+                   client_secret=client_secret, auth_type=SELF_DEPLOYED_AUTH_TYPE,
+                   scope=scope, resource=resource, app_url=app_url, *args, **kwargs)
+
+    def http_request(self, *args, **kwargs) -> requests.Response:
+        """
+        Overrides Base client request function, retrieves and adds to headers access token before sending the request.
+
+        Returns:
+            requests.Response: The http response
+        """
+        token = self.get_access_token()
+        headers = {
+            'Authorization': f'Bearer {token}',
+            'Content-Type': 'application/json',
+            'Accept': 'application/json'
+        }
+        return super()._http_request(*args, headers=headers, **kwargs)  # type: ignore[misc]
+
+    def get_access_token(self):
+        """
+        Obtains access and refresh token from oproxy server or just a token from a self deployed app.
+        Access token is used and stored in the integration context
+        until expiration time. After expiration, new refresh token and access token are obtained and stored in the
+        integration context.
+
+        Returns:
+            str: Access token that will be added to authorization header.
+        """
+        integration_context = demisto.getIntegrationContext()
+        access_token = integration_context.get('access_token')
+        valid_until = integration_context.get('valid_until')
+        if access_token and valid_until:
+            if self.epoch_seconds() < valid_until:
+                return access_token
+
+        auth_type = self.auth_type
+        refresh_token = ''
+        if auth_type == OPROXY_AUTH_TYPE:
+            access_token, expires_in, refresh_token = self._oproxy_authorize()
+        else:
+            access_token, expires_in = self._get_self_deployed_token()
```

Comment:
Refresh token is single use token that is should be sent to identity endpoint in order to receive access token and new refresh token. However `refresh_token` isn't being posted in `_get_self_deployed_token` method. In case of refresh token flow, the data dict should be of following form:
data = {
            'client_id': self.client_id,
            'client_secret': self.client_secret,
            'grant_type': **'refresh_token'**,
            'scope': self.scope,
            'redirect_uri': **redirect_uri**
}
For additional info (step 5): [Microsoft Docs](https://docs.microsoft.com/en-us/graph/auth-v2-user)
Additionally, you will need to add new field of `redirect_uri` in case of self deployed application mode.

Reply:
The self deployed flow is currently `client credentials`, it doesn't require a refresh token. Also, why do we need redirect URI?

Reply:
`client_credentials` flow requires admin consent?

Reply:
nope, it allows using the apps without any UI interaction.